File: ambeth/Ambeth.IoC.Test/ambeth/ioc/link/TestRegistry.cs
using De.Osthus.Ambeth.Ioc.Extendable;
using De.Osthus.Ambeth.Log;
using System.ComponentModel;

namespace De.Osthus.Ambeth.Ioc.Link
{
    public class TestRegistry : ITestListenerExtendable, IInitializingBean, ITestRegistry, INotifyPropertyChanged, IStartingBean
    {
        [LogInstance]
        public ILogger Log { private get; set; }

        protected readonly IExtendableContainer<PropertyChangedEventHandler> propertyChangedList = new DefaultExtendableContainer<PropertyChangedEventHandler>("pceListener");

        public event PropertyChangedEventHandler PropertyChanged
        {
            add
            {
                propertyChangedList.Register(value);
            }
            remove
            {
                propertyChangedList.Unregister(value);
            }
        }

        protected readonly IExtendableContainer<ITestListener> testListeners = new DefaultExtendableContainer<ITestListener>("testListener");

        public void AfterPropertiesSet()
        {
        }

        public void AfterStarted()
        {
            PropertyChangedEventArgs pceArgs = new PropertyChangedEventArgs("Test");
            PropertyChangedEventHandler[] pceListeners = propertyChangedList.GetExtensions();
            foreach (PropertyChangedEventHandler pceListener in pceListeners)
            {
        	    pceListener.Invoke(this, pceArgs);
            }
            ITestListener[] testListeners = this.testListeners.GetExtensions();
            foreach (ITestListener testListener in testListeners)
            {
        	    testListener.MyMethod(this, pceArgs);
            }
        }

        public PropertyChangedEventHandler[] GetPceListeners()
        {
            return propertyChangedList.GetExtensions();
        }

        public ITestListener[] GetTestListeners()
        {
            return testListeners.GetExtensions();
        }

        public void RegisterTestListener(ITestListener testListener)
        {
            testListeners.Register(testListener);
        }

        public void UnregisterTestListener(ITestListener testListener)
        {
            testListeners.Unregister(testListener);
        }
        
    }
}

File: jambeth/jambeth-ioc-test/src/test/java/de/osthus/ambeth/ioc/link/TestRegistry.java
package de.osthus.ambeth.ioc.link;

import java.beans.PropertyChangeEvent;
import java.beans.PropertyChangeListener;

import de.osthus.ambeth.ioc.DefaultExtendableContainer;
import de.osthus.ambeth.ioc.IInitializingBean;
import de.osthus.ambeth.ioc.IStartingBean;
import de.osthus.ambeth.ioc.extendable.IExtendableContainer;
import de.osthus.ambeth.log.ILogger;
import de.osthus.ambeth.log.LogInstance;

public class TestRegistry implements ITestListenerExtendable, IInitializingBean, ITestRegistry, IStartingBean
{
	@SuppressWarnings("unused")
	@LogInstance
	private ILogger log;

	protected final IExtendableContainer<PropertyChangeListener> propertyChangedList = new DefaultExtendableContainer<PropertyChangeListener>(
			PropertyChangeListener.class, "pceListener");

	protected final IExtendableContainer<ITestListener> testListeners = new DefaultExtendableContainer<ITestListener>(ITestListener.class, "testListener");

	@Override
	public void afterPropertiesSet() throws Throwable
	{
		// Intended blank
	}

	@Override
	public void afterStarted()
	{
		PropertyChangeEvent pceArgs = new PropertyChangeEvent(this, "Test", 1, 2);
		PropertyChangeListener[] pceListeners = propertyChangedList.getExtensions();
		for (PropertyChangeListener pceListener : pceListeners)
		{
			pceListener.propertyChange(pceArgs);
		}
		ITestListener[] testListeners = this.testListeners.getExtensions();
		for (ITestListener testListener : testListeners)
		{
			testListener.myMethod(pceArgs);
		}
	}

	@Override
	public PropertyChangeListener[] getPceListeners()
	{
		return propertyChangedList.getExtensions();
	}

	@Override
	public ITestListener[] getTestListeners()
	{
		return testListeners.getExtensions();
	}

	@Override
	public void registerTestListener(ITestListener testListener)
	{
		testListeners.register(testListener);
	}

	@Override
	public void unregisterTestListener(ITestListener testListener)
	{
		testListeners.unregister(testListener);
	}
}
